pkg/controller/podautoscaler/horizontal.go

```diff
@@ -703,52 +724,207 @@ func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.Horiz
 	return desiredReplicas
 }
 
-// convertDesiredReplicas performs the actual normalization, without depending on `HorizontalController` or `HorizontalPodAutoscaler`
-func convertDesiredReplicasWithRules(currentReplicas, desiredReplicas, hpaMinReplicas, hpaMaxReplicas int32) (int32, string, string) {
-
-	var minimumAllowedReplicas int32
-	var maximumAllowedReplicas int32
+// getScaleUpChangePerPeriod returns number of pods added during the constrate rate period
+func getScaleUpChangePerPeriod(scaleUpEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
+	period := time.Second * time.Duration(*args.ScaleUpConstraint.Rate.PeriodSeconds)
+	cutoff := time.Now().Add(-period)
+	var replicas int32
+	for _, rec := range scaleUpEvents[args.Key] {
+		if rec.timestamp.Before(cutoff) {
+			// outdated scale event should be erased for later usage
+			// We can't do it during storing events as at that point we don't know the PeriodSeconds
+			rec.outdated = true
+		} else {
+			replicas += rec.replicaChange
+		}
+	}
+	return replicas
+}
 
-	var possibleLimitingCondition string
-	var possibleLimitingReason string
+// getScaleDownChangePerPeriod returns number of pods removed during the constrate rate period
+func getScaleDownChangePerPeriod(scaleDownEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
+	period := time.Second * time.Duration(*args.ScaleUpConstraint.Rate.PeriodSeconds)
```

Comment:
Copy&paste bug has been introduced in ac5bb47c83e34cadaf86ca380a7048bb3b8d588f commit. Here is the trivial fix:
```suggestion
	period := time.Second * time.Duration(*args.ScaleDownConstraint.Rate.PeriodSeconds)
```

Reply:
Nice catch, thank you!
This is going to be a large PR. I haven't even started creating tests for the new functionality.